lib/langchain/messages/chat_message.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+module Langchain
+  module Messages
+    # Generic messages that can be used in the `chat` method of the LLM adapter
+    # It is recommended that `assistants/messages/*_messages.rb` be unified with this one in the future.
+    class ChatMessage
+      attr_reader :role, :content
+
+      ROLES = [:system, :user, :assistant].freeze
+
+      class << self
+        def system(content)
+          new(role: :system, content:)
+        end
+
+        def user(content)
+          new(role: :user, content:)
+        end
+
+        def assistant(content)
+          new(role: :assistant, content:)
+        end
+      end
+
+      def initialize(role:, content: "")
+        raise ArgumentError, "Role must be one of #{ROLES.join(", ")}" unless ROLES.member?(role.to_sym)
+
+        @role = role.to_sym
+        @content = content
+      end
+
+      def to_hash
+        {role: role.to_s, content:}
```

Comment:
How do you envision we would account for cases like [this](https://github.com/patterns-ai-core/langchainrb/blob/main/lib/langchain/assistants/messages/google_gemini_message.rb#L42-L61)?

Reply:
While I don't intend to go that far within the scope of this PR, if I  expand in the future, wouldn't it be desirable to split `ChatMessage` into two classes, such as `ChatMessage::Basic` and `ChatMessage::ToolCall`?
The former would be a simple class that only holds `role` and `content`, while the latter would be a class that additionally holds information necessary for tool calls.

By defining `to_hash` for both classes, they can be handled transparently.
For parts that cannot be fully handled by the default `to_hash`, each LLM adapter can absorb them. This way, I believe we can reduce the need for individual LLM support within `ChatMessage` and `Assistant`.

Reply:
> The former would be a simple class that only holds `role` and `content`, while the latter would be a class that additionally holds information necessary for tool calls.

Take a look at, for example, how tool results are sent to Anthropic (inside the `content` key): https://github.com/patterns-ai-core/langchainrb/blob/main/lib/langchain/assistants/messages/anthropic_message.rb#L32-L39

Reply:
I think for the purposes of this PR I would just collapse:
```
Langchain::Messages::AnthropicMessage
Langchain::Messages::GoogleGeminiMessage
Langchain::Messages::OpenAIMessage
```
into =>
```
class Langchain::ChatMessage
  def initialize()

  def to_openai_hash()

  def to_google_gemini_hash()

  def to_anthropic_hash()

  def to_hash()

end
```

What do you think? I can try take a stab at it!

Reply:
I was not very familiar with function calling, so I was surprised to find that the message formats are completely incompatible. Although I haven't been able to check all the documents, is my understanding below correct?

* In messages where the LLM instructs the use of a tool, the details of the tool to be called are specified in `tool_calls` (without using `content` or `tool_call_id`)
* In messages representing the result of the Assistant calling a tool as instructed, the call ID is stored in `tool_call_id`, and the result of the call is stored in `content` (without using `tool_calls`)
* While the above concepts are generally common among LLMs, the hash keys and hash formats differ

If this understanding is correct, couldn't we generalize it by creating classes like the following? (The namespace and implementation details of the classes are shown in a very simplified manner.)

```ruby
# Message for LLM to instruct the use of a tool
class ToolCalls < ChatMessage::Base
  # Single tool call
  class ToolCall < Data.new(:id, :tool_name, :method_name, :arguments)
  def initialize(tool_calls:)
    @tool_calls = tool_calls)
  end
  # ...
end

# Message representing the result of a tool call
class ToolCallResult < ChatMessage::Base
  def initialize(tool_call_id:, result:)
    @tool_call_id = tool_call_id
    @result = result
  end
  # ...
end
```

Since it seems difficult to generalize to_hash, how about performing the conversion for each LLM model as needed?

```ruby
class LLM::OpenAI
  def chat(params = {})
    if params[:messages].all? { *1.is*a?(Langchain::Messages::ChatMessage) }
      convert_chat_messages!(params)
    end
    # ...
  end

  def onvert_chat_messages!(params)
     params[:messages] = params[:messages].map | message |
        case message
        in ToolCalls(tool_calls:)
          tool_calls.map { tool_call_to_hash(_1) }
        in ToolCallResult(tool_call_id:, result:)
          { role: 'tool', tool_call_id:, content: result }
        else
          message.to_h
        end
    end
  end
end
```

Although the details need a bit more consideration, I am thinking of a policy to make the Messages structure as common as possible in this way and hide the differences among LLMs within each LLM class. What do you think?